crates/modelardb_server/src/data_folders.rs

```diff
@@ -0,0 +1,216 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Implementation of a struct that provides access to the local and remote data storage components.
+
+use std::sync::Arc;
+
+use deltalake_core::DeltaTableError;
+use modelardb_common::metadata::table_metadata_manager::TableMetadataManager;
+use modelardb_common::storage::DeltaLake;
+use modelardb_common::types::ServerMode;
+
+use crate::manager::Manager;
+use crate::ClusterMode;
+
+/// Folder for storing metadata and Apache Parquet files.
+#[derive(Clone)]
+pub struct DataFolder {
+    /// Delta Lake for storing metadata and Apache Parquet files.
+    pub delta_lake: Arc<DeltaLake>,
+    /// Metadata manager for providing access to metadata related to tables.
+    pub table_metadata_manager: Arc<TableMetadataManager>,
+}
+
+impl DataFolder {
+    /// Return a [`DataFolder`] created from `data_folder_path`. If the folder does not exist, it is
+    /// created. If the folder does not exist and could not be created or if the metadata tables could
+    /// not be created, [`DeltaTableError`] is returned.
+    pub async fn try_from_path(data_folder_path: &str) -> Result<Self, DeltaTableError> {
+        let delta_lake = DeltaLake::try_from_local_path(data_folder_path)?;
+        let table_metadata_manager = TableMetadataManager::try_from_path(data_folder_path).await?;
+
+        Ok(Self {
+            delta_lake: Arc::new(delta_lake),
+            table_metadata_manager: Arc::new(table_metadata_manager),
+        })
+    }
+
+    /// Return a [`DataFolder`] created from `connection_info`. If the connection information could not
+    /// be parsed or if the metadata tables could not be created, [`DeltaTableError`] is returned.
+    pub async fn try_from_connection_info(connection_info: &[u8]) -> Result<Self, DeltaTableError> {
+        let remote_delta_lake = DeltaLake::try_remote_from_connection_info(connection_info).await?;
+
+        let remote_table_metadata_manager =
+            TableMetadataManager::try_from_connection_info(connection_info).await?;
+
+        Ok(Self {
+            delta_lake: Arc::new(remote_delta_lake),
+            table_metadata_manager: Arc::new(remote_table_metadata_manager),
+        })
+    }
+}
+
+/// Folders for storing metadata and Apache Parquet files locally and remotely.
+#[derive(Clone)]
+pub struct DataFolders {
+    /// Folder for storing metadata and Apache Parquet files on the local file system.
+    pub local_data_folder: DataFolder,
+    /// Folder for storing metadata and Apache Parquet files in a remote object store.
+    pub maybe_remote_data_folder: Option<DataFolder>,
+    /// Folder from which metadata and Apache Parquet files will be read during query execution.
+    /// It is equivalent to `local_data_folder` when deployed on the edge and `remote_data_folder`
+    /// when deployed in the cloud.
+    pub query_data_folder: DataFolder,
```

Comment:
Can this not be a reference to the appropriate `DataFolder` instead of a copy? If so, maybe update the code.

Reply:
Since `DataFolder` consists of two `Arc` fields, I assume there is no reason to make it a reference since copying it will just copy the internal references.